src/cvsim/fit_curve.py

```diff
@@ -0,0 +1,192 @@
+        self.fixed_vars = {
+            'reduction_potential': reduction_potential,
+            'diffusion_reactant': diffusion_reactant,
+            'diffusion_product': diffusion_product,
+        }
+        self.fixed_vars = {k: v for k, v in self.fixed_vars.items() if v is not None}
+
+        # Values are [initial guess, lower bound, upper bound]
+        self.default_vars = {
+            'reduction_potential': [round((self.voltage_to_fit[np.argmax(self.current_to_fit)]
+                                           + self.voltage_to_fit[np.argmin(self.current_to_fit)]) / 2, 3),
+                                    round(min(self.start_voltage, self.reverse_voltage)/1000, 3),
+                                    round(max(self.start_voltage, self.reverse_voltage)/1000, 3),
+                                    ],
+            'diffusion_reactant': [1e-6, 5e-8, 1e-4],
+            'diffusion_product': [1e-6, 5e-8, 1e-4],
+        }
+
+    def fit(self,
+            reduction_potential: None | float | tuple[float, float] | tuple[float, float, float] = None,
+            diffusion_reactant: None | float | tuple[float, float] | tuple[float, float, float] = None,
+            diffusion_product: None | float | tuple[float, float] | tuple[float, float, float] = None,
+            ) -> tuple[np.ndarray, np.ndarray]:
+        """TODO  no guess | initial guess | bounds | guess, bounds"""
+
+        fit_vars = {
+            'reduction_potential': reduction_potential,
+            'diffusion_reactant': diffusion_reactant,
+            'diffusion_product': diffusion_product,
+        }
+        fit_vars = {k: v for k, v in fit_vars.items() if v is not None}
+
+        # check intersection of fixed_vars / fit_vars dicts. if so raise error
+        intersection_errors = (self.fixed_vars.keys() & fit_vars.keys())
+        if intersection_errors:
+            raise ValueError("Cannot input fixed value and guess value for same parameter")
+
+        # get params that will be fit
+        fitting_params = [
+            param for param in self.default_vars.keys()
+            if param not in self.fixed_vars.keys()
+        ]
+
+        # trim dict to set of fit variables
+        fit_default_vars = {k: v for k, v in self.default_vars.items() if k in fitting_params}
+        var_index = {var: index for index, var in enumerate(fit_default_vars.keys())}
+
+        # take fit_vars dict, for each in it, replace the initial guess/bounds if specified
+        for param, value in fit_vars.items():
+            if isinstance(value, float | int):
+                # Initial guess
+                fit_default_vars[param][0] = value
+            elif isinstance(value, tuple) and len(value) == 2:
+                # Lower and upper bound
+                fit_default_vars[param][1] = value[0]
+                fit_default_vars[param][2] = value[1]
+            elif isinstance(value, tuple) and len(value) == 3:
+                fit_default_vars[param] = value
```

Comment:
We might want some error messages for tuples of the wrong size too (e.g. 1 or > 3)

Reply:
done